src/gc/stringdedup.cpp

```diff
@@ -0,0 +1,155 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "stringdedup.h"
+#include "stringdedupqueue.h"
+#include "stringdedupthread.h"
+#include "stringdeduptable.h"
+#include "gcenv.h"
+
+    StringDedupThread* StringDedup::thread;   
+    StringDedupTable* StringDedup::table;   
+#ifdef MULTIPLE_HEAPS
+    size_t StringDedup::nqueues;
+    StringDedupQueue* StringDedup::queues;    
+    size_t* StringDedup::mem_write_sizes;
+    GCHashTableIteration* StringDedup::iters;
+#else
+    StringDedupQueue* StringDedup::queue; 
+    size_t StringDedup::mem_write_size;
+    GCHashTableIteration StringDedup::iter;
+#endif //MULTIPLE_HEAPS
+
+void StringDedup::Init(
+    size_t number_of_heaps
+)
+{
+    table = new StringDedupTable;
```

Comment:
Beware of allocation failures here. All of these will throw on failure and the GC is not allowed to throw. You should use the nothrow `operator new` and return a `bool` or something to indicate whether or not the initialization was successful.

It is a huge pain that we can't use `std::unique_ptr` here, and I'm sorry about that. I think that's something we definitely should address in the GC codebase as a whole going forward.